Fair tests balancing for integration tests

Felixoid opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Instead of balancing tests by `crc32(file_name)` we'll use `add tests to a group with a minimal number of tests`.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

Felixoid committed:
Add ClickhouseIntegrationTestsRunner.repo_path

Felixoid committed:
Sort pylint disables

Felixoid committed:
Distribute tests by batched jobs more equally

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #70272 from ClickHouse/ClickHouse:better-tests-balance -->
*This is an automated comment for commit 064f8c732811bf269e610cf1986d328ad483ea74 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70272/064f8c732811bf269e610cf1986d328ad483ea74/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/6474019a8483ebab01cb65defdd6b8626d062b77/builds/report.html">✅ success</a></td></tr>
<tr><td>ClickBench</td><td>Runs [ClickBench](https://github.com/ClickHouse/ClickBench/) with instant-attach table</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/clickbench__aarch64_.html">✅ success</a></td></tr>
<tr><td>Compatibility check</td><td>Checks that <b>clickhouse</b> binary runs on distributions with old libc versions. If it fails, ask a maintainer for help</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/compatibility_check__aarch64_.html">✅ success</a></td></tr>
<tr><td>Docker keeper image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/docker_keeper_image.html">✅ success</a></td></tr>
<tr><td>Docker server image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/docker_server_image.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/89d4ab65e6749864bf1aa7c587d6685a3e91269a/docs_check.html">✅ success</a></td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/fast_test.html">✅ success</a></td></tr>
<tr><td>Install packages</td><td>Checks that the built packages are installable in a clear environment</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/install_packages__aarch64_.html">✅ success</a></td></tr>
<tr><td>Performance Comparison</td><td>Measure changes in query performance. The performance test report is described in detail <a href="https://github.com/ClickHouse/ClickHouse/tree/master/docker/test/performance-comparison#how-to-read-the-report">here</a>. In square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/performance_comparison__release__[1_4]/report.html">✅ success</a></td></tr>
<tr><td>Stateful tests</td><td>Runs stateful functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/stateful_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Stateless tests</td><td>Runs stateless functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/stateless_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Stress test</td><td>Runs stateless functional tests concurrently from several clients to detect concurrency-related errors</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/stress_test__asan_.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/321980f6500daa523aaeb1bc5dc19b4536a71626/unit_tests__asan_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

Felixoid commented:
asan tests look distributed correctly:

```
1.  INFO:root:Found 517 tests first 3 test_MemoryTracking/test.py::test_http test_MemoryTracking/test.py::test_tcp_multiple_sessions test_MemoryTracking/test.py::test_tcp_single_session
2. INFO:root:Found 515 tests first 3 test_accept_invalid_certificate/test.py::test_accept test_accept_invalid_certificate/test.py::test_connection_accept test_accept_invalid_certificate/test.py::test_default
3. INFO:root:Found 512 tests first 3 test_access_control_on_cluster/test.py::test_access_control_on_cluster test_access_control_on_cluster/test.py::test_grant_all_on_cluster test_access_control_on_cluster/test.py::test_grant_current_database_on_cluster
4. INFO:root:Found 512 tests first 3 test_access_for_functions/test.py::test_access_rights_for_function test_access_for_functions/test.py::test_ignore_obsolete_grant_on_database test_alter_codec/test.py::test_alter_codec_index
5. INFO:root:Found 511 tests first 3 test_aggregation_memory_efficient/test.py::test_remote test_allowed_client_hosts/test.py::test_allowed_host test_allowed_client_hosts/test.py::test_denied_host
6.  INFO:root:Found 555 tests first 3 test_aliases_in_default_expr_not_break_table_structure/test.py::test_aliases_in_default_expr_not_break_table_structure[ReplicatedMergeTree] test_allowed_url_from_config/test.py::test_HDFS test_allowed_url_from_config/test.py::test_config_with_hosts
```

Felixoid committed:
Add a little bit of typing, fix logging

Felixoid committed:
Add cluster.wait_zookeeper_nodes_to_start

Both wait_zookeeper_secure_to_start and wait_zookeeper_to_start are the
same, and users creating wait_zookeeper_nodes_to_start here and there.

Felixoid committed:
Fix two simple flaky tests

Felixoid committed:
Slight typing improvements, add {replace,reset}_zookeeper_config

Felixoid committed:
Use common functions, fix flaky tests

Felixoid commented:
As promised, here are the fixes for https://s3.amazonaws.com/clickhouse-test-reports/70146/84c81cdc3cbdd7609d73b698266b8e0308824ea2/integration_tests_flaky_check__asan_.html from #70146
